examples/tutorials/imagenet_pytorch/model_def.py

```diff
@@ -0,0 +1,159 @@
+
+import torch
+import torch.nn as nn
+import torchvision.transforms as transforms
+import torchvision.datasets as datasets
+import torchvision.models as models
+
+
+TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]
+model_names = sorted(name for name in models.__dict__
+    if name.islower() and not name.startswith("__")
+    and callable(models.__dict__[name]))
+
+class ImagenetTrial(PyTorchTrial):
+    def __init__(self, context: PyTorchTrialContext):
+        self.context = context
+
+        arch = self.context.get_hparam("arch")
+        if self.context.get_hparam("pretrained"):
+            print("=> using pre-trained model '{}'".format(arch))
+            model = models.__dict__[arch](pretrained=True)
+        else:
+            print("=> creating model '{}'".format(arch))
+            model = models.__dict__[arch]()
+
+        self.model = self.context.wrap_model(model)
+
+        # Create a unique download directory for each rank so they don't overwrite each
+        # other when doing distributed training.
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.data_directory = self.context.get_hparam("data_location")
+
+        optimizer = torch.optim.SGD(self.model.parameters(), self.context.get_hparam("lr"), momentum=self.context.get_hparam("momentum"), weight_decay=self.context.get_hparam("weight_decay"))
+        self.optimizer = self.context.wrap_optimizer(optimizer)
+
+        self.criterion = nn.CrossEntropyLoss()
+
+        lr_sch = torch.optim.lr_scheduler.StepLR(self.optimizer, gamma=.1, step_size=2)
+        self.lr_sch = self.context.wrap_lr_scheduler(lr_sch, step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH)
+
+    def build_training_data_loader(self):
+
+        if self.context.get_hparam("dataset") == 'imagenet':
+            traindir = os.path.join(self.data_directory, 'train')
+            self.normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                        std=[0.229, 0.224, 0.225])
+
+            train_dataset = datasets.ImageFolder(
+                traindir,
+                transforms.Compose([
+                    transforms.RandomResizedCrop(224),
+                    transforms.RandomHorizontalFlip(),
+                    transforms.ToTensor(),
+                    self.normalize,
+                ]))
+
+            return DataLoader(
+                train_dataset, batch_size=self.context.get_per_slot_batch_size(), shuffle=True,
+                num_workers=self.context.get_hparam("workers", pin_memory=True))
+        
+        elif self.context.get_hparam('dataset') == 'cifar':
+
+            transform = transforms.Compose(
+                [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+            )
+            trainset = datasets.CIFAR10(
+                root=self.download_directory, train=True, download=True, transform=transform
```

Comment:
I'm not a big fan of hardcoding download=True here. I think it would be better to have that as a parameter in the config file that defaults to True just in case users want to change that and pass their own data directory. Ditto validation.

Reply:
This actually mimics our cifar example where we also set download == True. I've updated it though to at least be an hp flag